src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -364,6 +371,155 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                        // to ensure we're keeping lexical ordering of side effects.
+                        ensureTempTrackingSetup(ref temps, ref positionsAssignedToTemp);
+                        Debug.Assert(!positionsAssignedToTemp.IsNull);
+
+                        foreach (var placeholder in interpolationData.ArgumentPlaceholders)
+                        {
+                            // Replace each needed placeholder with a sequence of store and evaluate the temp.
+                            var argIndex = placeholder.ArgumentIndex;
+
+                            BoundLocal local;
+                            switch (argIndex)
+                            {
+                                case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter when receiverStored:
+                                    Debug.Assert(visitedReceiver != null && requiresInstanceReceiver);
+                                    local = (BoundLocal)((BoundSequence)visitedReceiver).Value;
+                                    break;
+
+                                case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter:
+                                    Debug.Assert(visitedReceiver != null && requiresInstanceReceiver);
+                                    local = _factory.StoreToTemp(visitedReceiver, out var store, refKind: visitedReceiver.GetRefKind());
+                                    temps.Add(local.LocalSymbol);
+                                    visitedReceiver = _factory.Sequence(ImmutableArray<LocalSymbol>.Empty, ImmutableArray.Create<BoundExpression>(store), local);
+                                    receiverStored = true;
+                                    break;
+
+                                case >= 0 when positionsAssignedToTemp[argIndex]:
+                                    local = visitedArgumentsBuilder[argIndex] switch
+                                    {
+                                        BoundSequence { Value: BoundLocal l } => l,
+                                        BoundLocal l => l, // Can happen for discard arguments
+                                        var u => throw ExceptionUtilities.UnexpectedValue(u)
+                                    };
+                                    break;
+
+                                case >= 0:
+                                    Debug.Assert(visitedArgumentsBuilder[argIndex] != null);
+                                    var paramIndex = argsToParamsOpt.IsDefault ? argIndex : argsToParamsOpt[argIndex];
+                                    RefKind argRefKind = argumentRefKindsOpt.RefKinds(argIndex);
+                                    RefKind paramRefKind = parameters[paramIndex].RefKind;
+                                    var visitedArgument = visitedArgumentsBuilder[argIndex];
+                                    local = _factory.StoreToTemp(visitedArgument, out store, refKind: paramRefKind == RefKind.In ? RefKind.In : argRefKind);
```

Comment:
Should we avoid creating extra temps if `visitedArgument` is already a temp?
Same question for `visitedReceiver` above (line 477).

Related question: what do we expect in terms of side-effects when using a `ref` argument and an argument attribute?
For instance, when calling `M(ref string x, [ArgumentAttribute("x")] CustomHandler handler)`, will `M` receive a value for `x` that is affected by the handler conversion, or are side-effects from the handler conversion supposed to be discarded?

Reply:
I don't think we should be trying to optimize temps in here. If we store an extra temp, so be it.

As to refs, refkinds must match. So if you pass something by ref, it can change values. There are tests covering this behavior.